Cardano streamed transaction signing

gabrielKerekes opened the pull request.

gabrielKerekes commented on src/js/core/methods/CardanoSignTransaction.js:
```diff
@@ -65,52 +94,29 @@ export default class CardanoSignTransaction extends AbstractMethod {
         ]);
 
-        const inputs: CardanoTxInputType[] = payload.inputs.map(input => {
+        const inputsWithPath: InputWithPath[] = payload.inputs.map(input => {
```
I need to separate `path` from the input since it's not getting passed into Trezor anymore.

gabrielKerekes commented on src/js/core/methods/CardanoSignTransaction.js:
```diff
@@ -130,21 +136,67 @@ export default class CardanoSignTransaction extends AbstractMethod {
+    }
+
+    _isFeatureSupported(feature: $Keys<typeof CardanoSignTransactionFeatures>) {
+        console.log({ v: this.device.getVersion() });
```
This will get removed.

gabrielKerekes commented on src/js/core/methods/CardanoSignTransaction.js:
```diff
@@ -155,38 +207,136 @@ export default class CardanoSignTransaction extends AbstractMethod {
+        };
+
+        // init
+        await typedCall('CardanoSignTxInit', 'CardanoTxItemAck', signTxInitMessage);
```
I wasn't able to think of a nicer way to send the messages one by one, but perhaps it can be done nicer?

gabrielKerekes commented on src/js/core/methods/helpers/cardanoCertificate.js:
```diff
@@ -122,38 +125,62 @@ const validatePoolParameters = (poolParameters: CardanoPoolParameters) => {
+            owners_count: poolParameters.owners.length,
+            relays_count: poolParameters.relays.length,
+        },
+        poolOwners: poolParameters.owners.map(owner => ({
```
Since `poolOwners` and `poolRelays` are sent in different messages I needed to separate them.

gabrielKerekes commented on src/js/core/methods/helpers/cardanoSignTxLegacy.js:
```diff
@@ -0,0 +1,121 @@
+    chain_code?: string,
+};
+
+export const toLegacyParams = params => {
```
I am not sure this is the best approach. Perhaps we could keep most of the old code in a separate file and decide earlier which version to use. In this approach, I first create the parameters for the new request and then look at FW version. If it's older, I use `toLegacyParams` to transform the parameters into the old format. This results in less duplication and less legacy code, but if the "new" params change, this transformation will maybe need to change also. Opinions?

refi93 replied:
hopefully we would be able to deprecate legacy mode as soon as the streamed one is out and people had a chance to update. If anything in the input format changes though and we would still need to keep it, we would very likely need to change whatever handles that legacy signing anyway

I'd cover the transformation with flow types at least though so the linter "screams" if any data structure changes (I see there's currently the `// @flow` annotation at the top missing btw, so I believe this file is left out from type checking)

gabrielKerekes commented on src/js/core/methods/helpers/cardanoSignTxLegacy.js:
```diff
@@ -0,0 +1,121 @@
+    }
+};
+
+export const serializedTxToResult = (txHash: string, serializedTx: string): CardanoSignedTxData => {
```
This I think is pretty much final.

gabrielKerekes commented on package.json:
```diff
@@ -106,6 +106,7 @@
     },
     "dependencies": {
         "@babel/runtime": "^7.12.5",
+        "borc": "^3.0.0",
```
I had to add a cbor library in order to parse the signed transactions of older FW.

refi93 replied:
Aren't the dependencies meant primarily for the connect outermost layer which resides on the web side (in case of browser integrations)?  I think the parsing would actually happen in the connect service, so this could be among extendedDependencies instead, not "bloating" further existing web integrations that have nothing to do with Cardano

refi93 commented on src/js/core/methods/helpers/cardanoCertificate.js:
```diff
@@ -122,38 +125,62 @@ const validatePoolParameters = (poolParameters: CardanoPoolParameters) => {
+            margin_numerator: poolParameters.margin.numerator,
+            margin_denominator: poolParameters.margin.denominator,
+            reward_account: poolParameters.rewardAccount,
+            owners: [],
```
does it make sense to keep owners and relays here if they are attached aside? Shouldn't there be the counts of owners and relays so Trezor knows how many to expect? If not, how does trezor know it from the protobuf message?

refi93 replied:
ah I see, the counts are below - perhaps I'd clarify with a comment the empty arrays are here just for wire compatibility with the legacy, non-streamed Cardano tx signing

gabrielKerekes committed:
feat(cardano): prepare protobuf messages and types for streamed tx signing

gabrielKerekes committed:
feat(cardano): streamed transaction signing

gabrielKerekes committed:
test(cardano): add tests for streamed transaction signing

szymonlesisz committed:
tests: add health check tests

- checking "extendedDependencies" in package.json
- checking "!fixup" commits

szymonlesisz committed:
chore: update protobuf messages

messages taken from trezor-firmware repo, commit 66bf309fbf98d3c3d8209285bf40e7e4a33c47cc
main changes: added "pages" and "page_number" fields in ButtonRequests

szymonlesisz committed:
feat: expose whole ButtonRequest object in UI.REQUEST_BUTTON event

szymonlesisz committed:
fix: TrezorConnect.dispose method

- added fetch AbortController (if available)  to bridge requests.
- implement dispose method in node and react-native indexes.
- remove eventListeners and references.
- dispose backend connections.

szymonlesisz committed:
fix: rename onBeforeUnload to dispose

vdovhanych committed:
feat(ci): automate production and beta releases to npm registry

marekrjpolak committed:
fix: TrezorConnect.firmwareUpdate method

- uploaded binary modified in the same way as when downloaded by version.

szymonlesisz committed:
chore: add 1.10.2 and 2.4.1 FW releases

szymonlesisz committed:
chore: update @trezor/rollout dependency v1.2.0

szymonlesisz committed:
fix: TrezorConnect.firmwareUpdate binary type

expected binary field is ArrayBuferr.
fix "paramsValidator" for binary object.
added unit test.

vdovhanych committed:
feat(ci): check connect version and commit message before merge to v8

marekrjpolak committed:
feat: add TrezorConnect.rebootToBootloader method
- currently works only for T1 with fw >= 1.10.0

tsusanka committed:
chore(changelog): add missing items for 8.1.29

vdovhanych committed:
feat(ci): add job for deploying connect to connect.trezor.io

vdovhanych committed:
fix(ci): fix publish beta release skipped state

vdovhanych committed:
fix(ci): fix publish beta release skipped state

vdovhanych committed:
fix(ci): fix release jobs to use nix-shell and resolve jobs being skipped

vdovhanych committed:
fix(ci): install all dependencies for yarn build

szymonlesisz committed:
tests: fix trezor-user-env loader

trezor-user-env websocket server was rewritten and returns different error

szymonlesisz committed:
tests: update getAccountInfo fixtures

do not rely on account balance, it may be changed

vdovhanych committed:
fix(ci): update configuration and fix release jobs

vdovhanych committed:
feat(ci): add beta check to publish npm beta job

vdovhanych committed:
release v8.1.30-beta.1

szymonlesisz committed:
fix(ci): build-extended suffix condition

fix of be3799f95d016647b18bfee0f44b929e9dc7b596

szymonlesisz committed:
fix: update missing version in runtime files

version was changed only in package.json be3799f95d016647b18bfee0f44b929e9dc7b596

szymonlesisz committed:
chore: move version bump scripts from Makefile to package.json

szymonlesisz committed:
chore: move build scripts from Makefile to package.json

szymonlesisz committed:
fix(ci): use build stage artifacts in production deploy

vdovhanych committed:
chore: update readme with a new deploy process and minor fixes

szymonlesisz committed:
fix(ci): update version-check script

- drop "compare-versions" dependency in favour of "semver"
- add check for "prerelease" versions with -beta suffix

vdovhanych committed:
fix(ci): run build and verify on deploy runner with nix shell

szymonlesisz committed:
chore: move protobuf script from Makefile to bash script

szymonlesisz committed:
chore: move coins script from Makefile to bash script

szymonlesisz committed:
chore: remove docker build from Makefile

- improved Dockerfile
- added possibility to build npm and npm-extended

szymonlesisz committed:
chore: move s3 sync job from Makefile to bash script

and remove Makefile completely

marekrjpolak committed:
fix: missing message for Failure_ActionCancelled

szymonlesisz committed:
release v8.1.30-beta.2

FrederikBolding committed:
chore: update trezor-common submodule

FrederikBolding committed:
feat: ethereumSignTransaction eip1559 support

szymonlesisz committed:
fix: getFirmwareRange find supported firmware rule from config.json

- override current range if min/max value equals "0"
- refactored search of "supportedFirmware" rule
- added unit tests for getFirmwareRange utility

szymonlesisz committed:
fix: device.unavailableCapabilities "no-support" condition

support should be determined by coins.json before reading capabilites from the device features

szymonlesisz committed:
chore: exclude __fixtures__ dir from npm build

szymonlesisz committed:
fix: rename config.json supportedFirmware rule field name (excludedMethods > methods)

szymonlesisz committed:
fix: distinguish between method and capability in coins.json

- methods: real trezor-connect methods validated in runtime using "getFirmwareRange" utility, not listed in device.unavailableCapabilities
- capabilities: defined feature potentially listed in device.unavailableCapabilities

gabrielKerekes committed:
chore(cardano): update docs and changelog

matejkriz committed:
docs: reflect new ETH transaction format introduced in EIP1559

szymonlesisz committed:
chore: remove Lisk related tests and docs

szymonlesisz committed:
chore: update trezor-common submodule [WIP]

szymonlesisz committed:
chore: remove Lisk related methods
